### DISCUSSION TITLE
fopen limited to 16 characters (including with path) [at least for cx16.lib]

### DISCUSSION THREAD

**Discussion by voidstar78** [cc65/cc65#2259] [General]
I understand CBM systems are historically limited to 16 character filenames (in total, including any extension).

But when using fopen, I might specify an absolute or relative path that can quickly go beyond 16 characters.   It appears that I can still specify an absolute or relative path, but it has to fit within that 16 characters.

Is this limitation a general cc65 thing for any platform (PET, apple, c64, etc?) or a cx16.lib limitation at the time it was produced?

I tried chopping off the path portion of the string myself and using chdir(x) but that didn't work either (the chdir compiled and ran, but never actually changed the directory).   I tried both style of slashes, or a relative path that didn't even need any slashes.  Does it work on c64 and other platforms and just not-yet-completed for CX16 platform?

Just curious if this is a known issue, or has something in the line of intermediate calls in fopen artificially limited the "filename" including path being limited to 16 characters?     Or otherwise, how do we open files in folders outside of the CWD in cc65?

**Comment by mrdudz** [comment]
This smells like an X/Y Problem....

however, the 16-Chars limit is imposed by the file system used by the CBM drives, the kernal or the bus protocol don't have such limits - and i don't think cc65 or it's library imposes any limits either (that wouldn't make a lot of sense in the first place)

**Comment by polluks** [comment]
Right now you can use a path with Apple ProDOS. If you have an
Ultimate, there is no UCI support so far. Feel free to implement a
change directory function yourself!

**Comment by polluks** [comment]
You are talking about CX16. Just send a CMDR-DOS command with cbm_open
channel 15 before you open the file.

**Comment by voidstar78** [comment]
Thanks, yes using the cbm_open/cbm_open(ch15)/cbm_read/cbm_cloes all seems to work with both long paths and long filenames, plus the flexibility of targeting a different device.    But I found that I have to check the channel 15 status even if I know that status will be "OK", otherwise the cbm_read might not actually work as expected (don't think that's a cc65 issue).

So to clarify, I had expected something like:
FILE* f = fopen("/stuff/reallyfar/away/overhere/somelongfilename.dat", "r");
to work.  And I understand it is not a C limitation or necessarily a cc65 limitation, but is on how the fopen is implemented for a given platform.  So the question was more, does such a long fopen call work on any of the standard target platforms? (I happen to be using CX16 which is a CBM target).

I found it wasn't an issue of supporting relative or absolute paths, I could still do short pathing like:
FILE* f = fopen("/abc/myfile.txt", "r");
Which is a little confusing, only in that it's not really that the filename is limited to 16 characters, but that "whatever you pass to fopen" is limited as such.


As suggested, I've switched over to the cbm_open/cbm_read routines, which are included "for free" with conio.h or separately in cbm.h     And that's fine, gives more control over the device and subchannel being used.  Just yes, whether the cbm_open result is 0 or not, one still needs to cbm_open ch15 to check the status before doing a reliable cbm_read.    


Was surprised chdir(x) linked for cx16 target.  fseek (or lseek) wasn't implemented and didn't even link, so felt if chdir is incomplete it shouldn't link either.   But I'm happy pressing on with cbm_xxx calls :)  Thanks!!

**Comment by mrdudz** [comment]
As said, (for cbm platforms) the limitations are imposed by the drive - you'll have to look at the drives DOS documentation to find out what it supports and what it doesn't support.

chdir() for that matter should work - it sends a cbm dos "cd" command on the cbm platforms. However the specific details again depend on the DOS itself.
